sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -62,11 +62,17 @@ trait ConstraintHelper {
    */
   def inferAdditionalConstraints(constraints: Set[Expression]): Set[Expression] = {
     var inferredConstraints = Set.empty[Expression]
-    constraints.foreach {
+    // IsNotNull should be constructed by `constructIsNotNullConstraints`.
+    val predicates = constraints.filterNot(_.isInstanceOf[IsNotNull])
+    predicates.foreach {
       case eq @ EqualTo(l: Attribute, r: Attribute) =>
-        val candidateConstraints = constraints - eq
+        val candidateConstraints = predicates - eq
         inferredConstraints ++= replaceConstraints(candidateConstraints, l, r)
         inferredConstraints ++= replaceConstraints(candidateConstraints, r, l)
+      case eq @ EqualTo(l @ Cast(_: Attribute, _, _), r: Attribute) =>
+        inferredConstraints ++= replaceConstraints(predicates - eq, r, l)
```

Comment:
according to https://github.com/apache/spark/pull/27252/files#r378111623

If we have `cast(a, ...) = b` and `b = 1`, we can definitely infer `cast(a, ...) = 1`.
If we have `cast(a, ...) = b` and `a = 1`, seems we can also infer `cast(1, ...) = b`.

But I'm a bit unsure about how to do it. We may need a variant of `replaceConstraints`, which holds an expression builder "e => cast(e, ...) = b". It looks for attribute `a`, and replace it with `cast(1, ...) = b`.

Reply:
I don't think we need to touch `replaceConstraints` at all. Please check https://github.com/apache/spark/pull/27518 that will do the trick because `a = 1` will be "substituted" into `cast(a, dt) = b` as a new `cast(1, dt) = b` constraint.

Reply:
I think both this and https://github.com/apache/spark/pull/27518 are beneficial. But I would use `val originalLeft = testRelation1.where('a < 1).subquery('left)` in test cases of this one instead of `val originalLeft = testRelation1.where('a === 1).subquery('left)` to avoid confusion.

Reply:
@cloud-fan This PR support `cast(1, dt) = b` before: 
https://github.com/apache/spark/compare/048a0ecc65763c6feaa939938e2dec6f4040d939..7dcfe915087dbe274b470928600197745a645f5e

I removed this because:
1. It may be broken the plan. [This is](https://github.com/apache/spark/compare/048a0ecc65763c6feaa939938e2dec6f4040d939..7dcfe915087dbe274b470928600197745a645f5e#diff-b40fcb6ac9b2e94b410f39a94a97e822R91-R93) how I handled it before.
2. For `cast(a, dt) = b`, we support inferring many predicates, for example: `a > 1`, `a < 1`, `a in (2, 3)`. I'm not sure if it's safe.

How about only supporting `cast(a, dt) = 1` now?

@peter-toth I'd like to support these cases in https://github.com/apache/spark/pull/27518:
```
a < b && b < c infer a < c
a < b && b <= c infer a < c
a < b && b = c infer a < c
...
```

Reply:
@wangyum I see, but I think currently you are doing something very different in https://github.com/apache/spark/pull/27518 see details here: https://github.com/apache/spark/pull/27518#issuecomment-585251432

Reply:
> How about only supporting cast(a, dt) = 1 now?

+1 for supporting the limited case only in this pr. Since this part of optimization can affect many queries, I think we need exhaustive discussions and tests for supporting wider cases.